feat(rust) dont store invalid credentials

crajcan opened the pull request:
<!-- Thank you for sending a pull request :heart: -->

## Current behavior

`ockam credential store` will store any string provided.

<!-- Please describe the current behavior of the code before the changes in this pull request is applied. -->

## Proposed changes
Verify just as `ockam credential verify` does.

https://github.com/build-trust/ockam/issues/4380
<!-- Please describe the changes proposed in this pull request. -->
<!-- If this pull request resolves an already recorded bug or a feature request, please add a link to that issue. -->

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

<!-- Looking forward to merging your contribution!! -->

crajcan committed:
fix(rust): test bad credentials not being stored

Test that invalid credentials are not stored by
`ockam credentials store` command

crajcan committed:
fix(rust): dont store invalid credentials

Verify credentials before storing in `credentials store`command
just as `credentials verify` does.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/credential/store.rs:
```diff
@@ -74,14 +76,37 @@ async fn run_impl(
+                *is_finished.lock().await = true;
+                Ok((false, format!("Credential is invalid\n{}", e.to_string())))
+            }
+        }
```
There are some simplifications you can do here. Note that you are returning a tuple with an ok/err pair. You could change that and return a plain Result<String> with the credential.

```rust
let issuer = match &cmd.identity().await {
    Ok(i) => i,
    Err(_) => {
        *is_finished.lock().await = true;
        return Err(miette!("Issuer is invalid").into());
    }
}
.identifier();
if validate_encoded_cred(&cred_as_str, &issuer, &vault_name, &opts)
    .await
    .is_err()
{
    *is_finished.lock().await = true;
    return Err(miette!("Credential is invalid").into());
};
Ok(cred_as_str)
```

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/credential/store.rs:
```diff
@@ -90,25 +115,29 @@ async fn run_impl(
+
+    if !is_valid {
+        Err(miette!(result).into())
+    } else {
```
If you apply the suggestion from the other comment, then this part also gets simplified:

```rust
let (credential, _) = try_join!(send_req, progress_output)?;
opts.terminal
    .stdout()
...
```

adrianbenavides commented on implementations/rust/ockam/ockam_command/tests/bats/credentials.bats:
```diff
@@ -42,3 +43,24 @@ teardown() {
   assert_output --partial "Attributes: {\"application\": \"Smart Factory\", \"city\": \"New York\""
 }
+
+@test "credential - verify and store reject invalid credentials" {
```
Love it, thanks for adding this test!

adrianbenavides commented on implementations/rust/ockam/ockam_command/tests/bats/credentials.bats:
```diff
@@ -42,3 +43,24 @@ teardown() {
+  run "$OCKAM" credential show smart_la_cred
+  assert_failure
+  assert_output --partial "Unable to find credential named smart_la_cred"
+}
```
Make sure to run the `shfmt` tool to fix any formatting issues. [Check this out](https://github.com/build-trust/ockam/blob/a83c8e9c8dc7b90221dcc0b4cacd45356bda9f98/implementations/rust/ockam/ockam_command/tests/bats/BATS.md#how-to-format-the-tests-scripts).

adrianbenavides reviewed (commented):
Thank you so much @crajcan for submitting this PR, it looks awesome! I've left some minor comments with a couple of thins to fix.

mrinalwadhwa commented:
@crajcan Thank you for spending time on this, and an awesome first PR. Before we can merge please

- [ ] Accept the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding your Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will make the commit as verified.

crajcan committed:
fix(rust): address pr comments and lint

Apply suggest refactors. Run clippy, fmt, shfmt

davide-baldo commented:
Thank you for taking the time to contribute to our project. Your work is valuable, and we would love to see it through to completion. How can we help you in finishing it?

crajcan commented:
> Thank you for taking the time to contribute to our project. Your work is valuable, and we would love to see it through to completion. How can we help you in finishing it?

My fault, I thought it was just awaiting approval. I just pushed an attempt to fix the lints, I can look at the failing test this afternoon.

crajcan committed:
test(rust): fix trust context test

Since the store command no longer stores invalid credentials, trust contexts cannot be created
with invalid credentials either.

davide-baldo reviewed (approved):
Thank you for your contribution!
